langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Sort.java

```diff
@@ -18,73 +18,185 @@
-            } else if (comparator.call(new Object[]{strand, aux.get(j), true, aux.get(i), true}) < 0) {
-                input.add(k, aux.get(j++));
+                index = i++;
+            } else if (isAscending && sortFunc(aux[j][0], aux[i][0], type, true) < 0) {
+                index = j++;
+            } else if (!isAscending && sortFunc(aux[i][0], aux[j][0], type, false) < 0) {
+                index = j++;
             } else {
-                input.add(k, aux.get(i++));
+                index = i++;
             }
+            input[k] = aux[index];
         }
     }
 
+    private static int sortFunc(Object value1, Object value2, BType type, boolean isAscending) {
+        int c = -999;
+        // () should come last irrespective of the sort direction.
+        if (value1 == null) {
+            if (value2 == null) {
+                return 0;
+            }
+            if (isAscending) {
+                return 1;
+            }
+            return -1;
+        } else if (value2 == null) {
+            if (isAscending) {
+                return -1;
+            }
+            return 1;
+        } else if (type.getTag() == TypeTags.INT_TAG) {
+            c = Long.compare((long) value1, (long) value2);
+        } else if (type.getTag() == TypeTags.FLOAT_TAG) {
+            // NaN should be placed last or one before the last when () is present irrespective of the sort direction.
+            if (Double.isNaN((double) value1)) {
+                if (Double.isNaN((double) value2)) {
+                    return 0;
+                }
+                if (isAscending) {
+                    return 1;
+                }
+                return -1;
+            } else if (Double.isNaN((double) value2)) {
+                if (isAscending) {
+                    return -1;
+                }
+                return 1;
+            }
+            c = Double.compare((double) value1, (double) value2);
+        } else if (type.getTag() == TypeTags.DECIMAL_TAG) {
+            c = new BigDecimal(value1.toString()).compareTo(new BigDecimal(value2.toString()));
+        } else if (type.getTag() == TypeTags.BOOLEAN_TAG) {
+            c = Boolean.compare((boolean) value1, (boolean) value2);
+        } else if (type.getTag() == TypeTags.STRING_TAG) {
+            c = value1.toString().compareToIgnoreCase(value2.toString());
+        } else if (type.getTag() == TypeTags.BYTE_TAG) {
+            c = Integer.compare((int) value1, (int) value2);
+        } else if (type.getTag() == TypeTags.ARRAY_TAG) {
+            for (int i = 0; i < ((ArrayValue) value1).size(); i++) {
+                c = sortFunc(((ArrayValue) value1).get(i), ((ArrayValue) value2).get(i),
+                        ((ArrayValue) value1).getElementType(), isAscending);
+                if (c != 0) {
+                    break;
+                }
+            }
+        } else {
+            throw BallerinaErrors.createError("invalid sort key type", "Failed to sort array");
```

Comment:
This doesn't seem to follow the normal practice we follow for langlib errors. We currently have a simple string as the error message and the second arg goes into the detail. Please check other usage of this method.